src/contracts/sgho/sGHO.sol

```diff
@@ -0,0 +1,380 @@
+    }
+  }
+
+  /**
+   * @notice Converts a GHO asset amount to a sGHO share amount based on the current yield index.
+   * @dev Overrides the standard ERC4626 implementation to use the custom yield-based conversion.
+   * @param assets The amount of GHO assets.
+   * @param rounding The rounding direction to use.
+   * @return The corresponding amount of sGHO shares.
+   */
+  function _convertToShares(
+    uint256 assets,
+    Math.Rounding rounding
+  ) internal view virtual override returns (uint256) {
+    uint256 currentYieldIndex = _getCurrentYieldIndex();
+    if (currentYieldIndex == 0) return 0;
+    return assets.mulDiv(RAY, currentYieldIndex, rounding);
+  }
+
+  /**
+   * @notice Converts a sGHO share amount to a GHO asset amount based on the current yield index.
+   * @dev Overrides the standard ERC4626 implementation to use the custom yield-based conversion.
+   * @param shares The amount of sGHO shares.
+   * @param rounding The rounding direction to use.
+   * @return The corresponding amount of GHO assets.
+   */
+  function _convertToAssets(
+    uint256 shares,
+    Math.Rounding rounding
+  ) internal view virtual override returns (uint256) {
+    uint256 currentYieldIndex = _getCurrentYieldIndex();
+    return shares.mulDiv(currentYieldIndex, RAY, rounding);
+  }
+
+  /**
+   * @notice Calculates the current yield index, including yield accrued since the last update.
+   * @dev This is a view function and does not modify state. It's used for previews.
+   * The interest calculation is linear within each update period, but compounds across multiple updates.
+   * Formula: newIndex = oldIndex * (1 + rate * time)
+   * Uses SafeCast to prevent overflow when casting to uint176. If overflow occurs, the transaction will revert
+   * instead of silently wrapping, protecting user rewards.
+   * @return The current yield index.
+   */
+  function _getCurrentYieldIndex() internal view returns (uint176) {
+    sGHOStorage storage $ = _getsGHOStorage();
+    if ($.ratePerSecond == 0) return $.yieldIndex;
+
+    uint256 timeSinceLastUpdate = block.timestamp - $.lastUpdate;
+    if (timeSinceLastUpdate == 0) return $.yieldIndex;
+
+    // Linear interest calculation for this update period: newIndex = oldIndex * (1 + rate * time)
+    // True compounding occurs through multiple updates as each update builds on the previous index
+    uint256 accumulatedRate = $.ratePerSecond * timeSinceLastUpdate;
+    uint256 growthFactor = RAY + accumulatedRate;
+
+    return (($.yieldIndex * growthFactor) / RAY).toUint176();
+  }
+
+  /**
+   * @notice Updates the yield index to accrue yield up to the current timestamp.
+   * @dev This function modifies state and is called before any operation that depends on the yield index.
+   * Uses SafeCast to prevent overflow when casting to uint176. If overflow occurs, the transaction will revert
+   * instead of silently wrapping, protecting user rewards.
+   */
+  function _updateYieldIndex() internal {
```

Comment:
please follow Solidity style guidelines: https://docs.soliditylang.org/en/latest/style-guide.html#order-of-layout

> Contract elements should be laid out in the following order:
> Pragma statements
> Import statements
> Events
> Errors
> Interfaces
> Libraries
> Contracts
> Inside each contract, library or interface, use the following order:
> Type declarations
> State variables
> Events
> Errors
> Modifiers
> Functions
>